publisher/publisher.go

```diff
@@ -0,0 +1,264 @@
+	logDesc.ID = base64.StdEncoding.EncodeToString(pkHash[:])
+	if len(logDesc.ID) != 44 {
+		return fmt.Errorf("Invalid log ID length [%d]", len(logDesc.ID))
+	}
+
+	return nil
+}
+
+// CTConfig defines the JSON configuration file schema
+type CTConfig struct {
+	Logs              []LogDescription `json:"logs"`
+	SubmissionRetries int              `json:"submissionRetries"`
+	// This should use the same method as the DNS resolver
+	SubmissionBackoffString string        `json:"submissionBackoff"`
+	SubmissionBackoff       time.Duration `json:"-"`
+
+	BundleFilename string   `json:"intermediateBundleFilename"`
+	IssuerBundle   []string `json:"-"`
+}
+
+type ctSubmissionRequest struct {
+	Chain []string `json:"chain"`
+}
+
+const (
+	sctVersion       = 0
+	sctSigType       = 0
+	sctX509EntryType = 0
+)
+
+// PublisherImpl defines a Publisher
+type PublisherImpl struct {
+	log               *blog.AuditLogger
+	client            *http.Client
+	submissionBackoff time.Duration
+	submissionRetries int
+	issuerBundle      []string
+	ctLogs            []LogDescription
+
+	SA core.StorageAuthority
+}
+
+// NewPublisherImpl creates a Publisher that will submit certificates
+// to any CT logs configured in CTConfig
+func NewPublisherImpl(ctConfig CTConfig) (pub PublisherImpl, err error) {
+	logger := blog.GetAuditLogger()
+	logger.Notice("Publisher Authority Starting")
+
+	if ctConfig.BundleFilename == "" {
+		err = fmt.Errorf("No CT submission bundle provided")
+		return
+	}
+	bundle, err := core.LoadCertBundle(ctConfig.BundleFilename)
+	if err != nil {
+		return
+	}
+	for _, cert := range bundle {
+		pub.issuerBundle = append(pub.issuerBundle, base64.StdEncoding.EncodeToString(cert.Raw))
+	}
+	ctBackoff, err := time.ParseDuration(ctConfig.SubmissionBackoffString)
+	if err != nil {
+		return
+	}
+	for _, log := range ctConfig.Logs {
+		if !strings.HasPrefix(log.URI, "https://") && !strings.HasPrefix(log.URI, "http://") {
+			err = fmt.Errorf("Log URI [%s] is not absolute", log.URI)
+			return
+		}
+	}
+
+	pub.log = logger
+	pub.client = &http.Client{}
+	pub.submissionBackoff = ctBackoff
+	pub.submissionRetries = ctConfig.SubmissionRetries
+	pub.ctLogs = ctConfig.Logs
+
+	return
+}
+
+func (pub *PublisherImpl) submitToCTLog(serial string, jsonSubmission []byte, log LogDescription) error {
+	done := false
+	var sct core.SignedCertificateTimestamp
+	backoff := pub.submissionBackoff
+	var retries int
+	for retries = 0; retries <= pub.submissionRetries; retries++ {
+		if retries > 0 {
+			time.Sleep(backoff)
+		}
+		resp, err := postJSON(pub.client, fmt.Sprintf("%s%s", log.URI, "/ct/v1/add-chain"), jsonSubmission, &sct)
+		if err != nil {
+			// Retry the request, log the error
+			pub.log.Warning(fmt.Sprintf("Error POSTing JSON to CT log submission endpoint [%s]: %s", log.URI, err))
+			backoff = pub.submissionBackoff
+			continue
+		} else if resp.StatusCode == http.StatusRequestTimeout || resp.StatusCode == http.StatusServiceUnavailable {
```

Comment:
I think other status codes, e.g. StatusForbidden will fall through to the done = true clause, right? I think we need one more `else if resp.StatusCode != http.StatusOK`. Probably any status codes other than RequestTimeout, ServiceUnavailable, and OK should not be retried: we should just return immediately with an error.